ibtk/src/utilities/MarkerPatchHierarchy.cpp

```diff
@@ -380,6 +753,49 @@ MarkerPatchHierarchy::setVelocities(const int u_idx, const std::string& kernel)
     }
 }
 
+void
+MarkerPatchHierarchy::forwardEulerStep(const double dt, const int u_new_idx, const std::string& kernel)
+{
+    const auto rank = IBTK_MPI::getRank();
+    for (int ln = d_hierarchy->getFinestLevelNumber(); ln >= 0; --ln)
+    {
+        unsigned int local_patch_num = 0;
+        Pointer<PatchLevel<NDIM> > current_level = d_hierarchy->getPatchLevel(ln);
+        for (int p = 0; p < current_level->getNumberOfPatches(); ++p)
+        {
+            if (rank == current_level->getMappingForPatch(p))
+            {
+                MarkerPatch& marker_patch = d_marker_patches[ln][local_patch_num];
+                Pointer<Patch<NDIM> > patch = current_level->getPatch(p);
+
+                // 1. Do a forward Euler step:
+                std::vector<double> new_positions(marker_patch.d_positions);
+                for (unsigned int i = 0; i < marker_patch.d_positions.size(); ++i)
+                {
+                    new_positions[i] = marker_patch.d_positions[i] + dt * marker_patch.d_velocities[i];
+                }
+
+                // 2. Interpolate the velocity at the new time:
+                std::vector<double> new_velocities(marker_patch.d_velocities);
+                do_interpolation(u_new_idx, new_positions, patch, kernel, new_velocities);
+
+                // 3. Do a backward Euler step:
```

Comment:
I had to look up that this is what `FORWARD_EULER` time stepping does for particles with `IBExplicitHierarchyIntegrator`. Seems like I might have missed adding a `break` statement in one of those switch statements. 🤷‍♂️

Speaking of --- is there a reason not to follow the pattern in `IBExplicitHierarchyIntegrator` and pass in the time stepping type and have a common time stepping routine? I guess one reason is that it makes it easier to see through the interface what time stepping is provided...

Reply:
> I had to look up that this is what FORWARD_EULER time stepping does for particles with IBExplicitHierarchyIntegrator. Seems like I might have missed adding a break statement in one of those switch statements. :man_shrugging:

Yeah, I wasn't sure what was going on with that one but I think it's still a convergent time integrator? Should we fix that? I have no idea if anyone is actually relying on that time integrator right now.

> s there a reason not to follow the pattern in IBExplicitHierarchyIntegrator and pass in the time stepping type and have a common time stepping routine?

We could, but it is slightly simpler to defer the choice of timestepping to the caller. I'd rather not carry the extra state around.

Interestingly, since markers have a one-way coupling with the velocity field, all the time integrators are actually equivalent - the velocity field is linear in time so midpoint/trapezoid/whatever forward Euler turns out to be are all equal. If we use the same time integrator we get exactly the same coordinates for the markers as IB points. If we use a different one then it is off by roundoff.